src/DotNetty.Buffers/ByteBufferUtil.cs

```diff
@@ -569,5 +569,12 @@ static void AppendHexDumpRowPrefix(StringBuilder dump, int row, int rowStartInde
                 dump.Append('|');
             }
         }
+
+        public static string DecodeString(IByteBuffer src, int readerIndex, int len, Encoding encoding)
+        {
+            if (len == 0) return string.Empty;
+            
+            return encoding.GetString(src.Array.Slice(len));
```

Comment:
To clarify, with distinction based on `HasArray` `DecodeString` would still match the behavior of netty in generic way that is also most efficient for the common case (byte array backed buffer). There's no need to implement special ToString on CompositeByteBuffer at this point.